src/compute-client/src/command.rs

```diff
@@ -508,14 +515,19 @@ impl<T: timely::progress::Timestamp> ComputeCommandHistory<T> {
         // First determine what the final compacted frontiers will be for each collection.
         // These will determine for each collection whether the command that creates it is required,
         // and if required what `as_of` frontier should be used for its updated command.
-        let mut final_frontiers = std::collections::BTreeMap::new();
+        let mut final_frontiers = BTreeMap::new();
         let mut live_dataflows = Vec::new();
         let mut live_peeks = Vec::new();
-        let mut live_cancels = std::collections::BTreeSet::new();
+        let mut live_cancels = BTreeSet::new();
 
         let mut create_inst_command = None;
         let mut create_timely_command = None;
-        let mut update_max_result_size_command = None;
+
+        // Collect only the final configuration.
+        // Note that this is only correct as long as all config parameters apply globally. If we
+        // ever introduce parameters that only affect subsequent commands, we will have to
+        // reconsider this approach.
```

Comment:
I am a bit confused about this comment. Is it not the case that if we update the configuration now in a normal run, then the updated configuration (e.g., smaller maximum result size) would apply to the subsequent commands only?

Reply:
No, it also applies to pending peeks that have not yet been fulfilled by the replica, as well as to subsequent batches emitted by in-progress subscribes. Updating the max result size simply updates the `max_result_size` field in the `ComputeState`. This field is then read every time:

* We check pending peeks: https://github.com/MaterializeInc/materialize/blob/dad83099a1cdd015c8cf9d11ee2a9671561971e2/src/compute/src/compute_state.rs#L657
* We emit results from in-progress subscribes: https://github.com/MaterializeInc/materialize/blob/dad83099a1cdd015c8cf9d11ee2a9671561971e2/src/compute/src/compute_state.rs#L692

This includes peeks and subscribes that were created before the max result size was updated.

Reply:
I see... so our current approach can make the fulfillment of peeks be non-deterministic wrt. `UpdateConfiguration` commands, right? Suppose we have a peek that should return a size of 100 and a single replica. The history has an `UpdateConfiguration(max_result_size = 10)` followed by the referred `Peek` (which goes pending) followed by zero or more commands followed by another `UpdateConfiguration(max_result_size = 1000)`. The fulfillment of the peek races with the update to the configuration. So in some executions this history fragment might produce a peek fulfillment and in some executions it might not. However, if the replica fails before the peek fulfillment, but after the controller received the last `UpdateConfiguration` command, and we reduce the history upon restart, then we will certainly fulfill the peek.

A scenario such as the one above does not sound too scary to me in the case of peeks/subscribes, so I am happy to go ahead with the current change. But we might need to rethink things if any configuration change could race with state we'd not like to diverge across replicas...